src/compose/compose-collection.ts

```diff
@@ -16,68 +16,132 @@ import { resolveBlockMap } from './resolve-block-map.js'
+      : token.type === 'block-seq'
+      ? resolveBlockSeq(CN, ctx, token, onError, tag)
+      : resolveFlowCollection(CN, ctx, token, onError, tag)
+
+  const Coll = coll.constructor as typeof YAMLMap | typeof YAMLSeq
+
+  // If we got a tagName matching the class, or the tag name is '!',
+  // then use the tagName from the node class used to create it.
+  if (tagName === '!' || tagName === Coll.tagName) {
+    coll.tag = Coll.tagName
+    return coll
+  } else if (tagName) {
+    coll.tag = tagName
+  }
+
+  return coll
+}
+
 export function composeCollection(
   CN: ComposeNode,
   ctx: ComposeContext,
   token: BlockMap | BlockSequence | FlowCollection,
   tagToken: SourceToken | null,
   onError: ComposeErrorHandler
 ) {
+  const tagName: string | null = !tagToken
+    ? null
+    : ctx.directives.tagName(tagToken.source, msg =>
+        onError(tagToken, 'TAG_RESOLVE_FAILED', msg)
+      )
+
   let coll: YAMLMap.Parsed | YAMLSeq.Parsed
-  switch (token.type) {
-    case 'block-map': {
-      coll = resolveBlockMap(CN, ctx, token, onError)
-      break
-    }
-    case 'block-seq': {
-      coll = resolveBlockSeq(CN, ctx, token, onError)
-      break
-    }
-    case 'flow-collection': {
-      coll = resolveFlowCollection(CN, ctx, token, onError)
-      break
-    }
+  let expType: 'map' | 'seq' | undefined =
+    token.type === 'block-map'
+      ? 'map'
+      : token.type === 'block-seq'
+      ? 'seq'
+      : token.type === 'flow-collection'
+      ? token.start.source === '{'
+        ? 'map'
+        : 'seq'
+      : undefined
+
+  if (!expType) {
+    onError(
+      token,
+      'IMPOSSIBLE',
+      'could not determine collection expression type',
+      true
+    )
   }
 
-  if (!tagToken) return coll
-  const tagName = ctx.directives.tagName(tagToken.source, msg =>
-    onError(tagToken, 'TAG_RESOLVE_FAILED', msg)
-  )
-  if (!tagName) return coll
+  // shortcut: check if it's a generic YAMLMap or YAMLSeq
+  // before jumping into the custom tag logic.
+  if (
+    !tagToken ||
+    !tagName ||
+    tagName === '!' ||
+    (tagName === YAMLMap.tagName && expType === 'map') ||
+    (tagName === YAMLSeq.tagName && expType === 'seq') ||
+    !expType
+  ) {
+    return resolveCollection(CN, ctx, token, onError, tagName)
+  }
 
-  // Cast needed due to: https://github.com/Microsoft/TypeScript/issues/3841
-  const Coll = coll.constructor as typeof YAMLMap | typeof YAMLSeq
-  if (tagName === '!' || tagName === Coll.tagName) {
-    coll.tag = Coll.tagName
-    return coll
+  let tag = ctx.schema.tags.find(t => t.tag === tagName) as
+    | CollectionTag
+    | undefined
+  if (tag && tag.collection !== expType) {
+    if (tag.collection) {
+      onError(
+        tagToken,
+        'BAD_COLLECTION_TYPE',
+        `${tag.tag} used for ${expType} collection, but expects ${tag.collection}`,
+        true
+      )
+      return resolveCollection(CN, ctx, token, onError, tagName)
+    }
+    tag = undefined
```

Comment:
Is that true?  Because previously it was only selecting the tag if both the tag matched and the `tag.collection === exptype`.

I don't have any ready examples of a custom tag that might be able to do both maps and sequences, but yes, if it's possible now and not with this change, then that is technically a breaking change, and we should probably figure out how to keep supporting it.

Reply:
The way you can do it is to have two separate entries in `customTags` with the same `tag` name but different `collection` types. I don't really know _why_ you'd want that, but it's technically possible with the current API.

For scalars, multiple tag entries are used to allow for different [number representations](https://github.com/eemeli/yaml/blob/main/src/schema/core/int.ts) to be maintained. So it's possible someone's come up with a reason to do something similar with collections.

Reply:
Ah, ok, then. I was imagining it might be nice to explicitly indicate that a tag _can't_ handle a given collection type, but I suppose it still errors in that case if the selection is tagName && expType, just with a "can't resolve tag" error instead of a "tag can't handle this kind of thing" error.  I'll refactor it back into that shape.